k4FWCore/include/k4FWCore/FunctionalUtils.h

```diff
@@ -133,16 +136,34 @@ namespace k4FWCore {
           using EDM4hepType = std::remove_cv_t<std::remove_pointer_t<typename TupleType::value_type>>;
           auto inputMap     = std::vector<const EDM4hepType*>();
           for (auto& handle : std::get<Index>(handles)) {
-            podio::CollectionBase* in = handle.get()->get();
-            inputMap.push_back(static_cast<const EDM4hepType*>(in));
+            podio::CollectionBase* in      = handle.get()->get();
+            auto*                  typedIn = dynamic_cast<const EDM4hepType*>(in);
+            if (typedIn) {
+              inputMap.push_back(typedIn);
+            } else {
+              throw GaudiException(
+                  thisClass->name(),
+                  fmt::format("Failed to cast collection {} to the required type {}, the type of the collection is {}",
+                              handle.objKey(), typeid(EDM4hepType).name(), in ? in->getTypeName() : "[undetermined]"),
```

Comment:
Because of the if - else that was removed in #276 `typeid` has to be used here... I think it's acceptable considering this will only happen for when reading using an `std::vector` unlike the usual normal input below. It looks much cleaner without and if - else to check if it's a `podio::CollectionBase`.

Reply:
Yeah, I think it's OK. If people really need more we can always make it better later.